hack/generator/spec/case-study-fixed-storage-version.md

```diff
@@ -0,0 +1,1105 @@
+
+The generated `ConvertToStorage()` and `ConvertFromStorage()` methods will test for the presence of this interface and will call it if available:
+
+``` go
+package v20140404preview
+
+import v1
+
+// ConvertTo converts this Person to the Hub version.
+func (person *Person) ConvertToStorage(dest v1.Person) error {
+    // … elided …
+
+    // *** Check for the interface and use it if found ***
+    if assignable, ok := person.(AssignableToPerson); ok {
+        assignable.AssignTo(dest)
+    }
+
+    return nil
+}
+
+// ConvertFrom converts from the Hub version to this version.
+func (person *Person) ConvertFromStorage(source v1.Person) error {
+    // … elided …
+
+    // *** Check for the interface and use it if found ***
+    if assignable, ok := person.(AssignableFromPerson); ok {
+        assignable.AssignFrom(source)
+    }
+
+    return nil
+}
+```
+
+Implementations of these interfaces are called *after* the generated boilerplate conversion has completed. This gives developers freedom to make any changes required as they won't be overwritted by the effects of generated code.
+
+## Version Map
+
+The preview version just appears as another version in our hub and spoke diagram:
+
+![](images/case-study-fixed-storage-2014-04-04-preview.png)
+
+
+# Version 2014-04-04 - Schema Change
+
+Based on feedback generated by the preview release, the CRM schema changes have gone ahead with a few minor changes:
+
+``` go
+package v20140404
+
+type Person struct {
+    Id         Guid
+    LegalName  string // *** Was FullName in preview ***
+    FamilyName string
+    KnownAs    string
+    AlphaKey   string // *** Added after preview ***
+}
+```
+
+The two new properties need to be added to the storage version to allow this to be stored:
+
+``` go
+package v1
+
+type Person struct {
+    Id         *Guid
+    AlphaKey   *string // *** New ***
+    FamilyName *string
+    FirstName  *string
+    FullName   *string
+    KnownAs    *string
+    LastName   *string
+    LegalName  *string // *** New ***
+    MiddleName *string
+}
+
+// Hub marks this type as a conversion hub.
+func (*Person) Hub() {}
+```
+
+## Issue: Property Bloat
+
+As our API evolves over time, our storage version is accumulating all the properties that have ever existed, bloating the storage version with obsolete properties that are seldom (if ever) used. 
+
+For example, only preview users would ever have used `FullName` as it became `LegalName` in the general release; most users will never use `FullName`.
+
+We have a problem, however. We can't remove the `FullName` property as that is a breaking change that will negatively impact users who *have* used the preview version. Both properties will need to be retained permanently.
+
+This violates the *pay for play* principle - even users who adopt the operator after the 2014-04-04 release will have to deal with the complexity even though they've never used the preview version.
```

Comment:
The *users* wouldn't have to deal with any additional complexity, would they?

Reply:
On the happy path, when everything works, no. In that case, nobody cares about the format.

But when something goes awry and is investigating issues, they may choose (or we may ask them) to directly inspect what's in etcd to work out what's going on; the more complex the information that's returned, the harder it becomes to troubleshoot the issues.